File: runtime/c/ledger.c
#include "ledger.h"

#include "data.h"
#include "store.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

struct LanaLedger {
    LanaStore *store;
    FILE *journal;
};

static bool text_ok(const char *text) { return text != NULL && text[0] != '\0'; }

static LanaError map_text(const LanaMap *map, const char *key, const char **out) {
    Value value;
    if (lana_map_get(map, key, &value) != LANA_OK || value.type != VAL_STRING || !text_ok(value.as.string)) return LANA_ERR_CORRUPTION;
    *out = value.as.string; return LANA_OK;
}

static LanaError map_number(const LanaMap *map, const char *key, uint64_t *out) {
    Value value;
    if (lana_map_get(map, key, &value) != LANA_OK || value.type != VAL_NUMBER) return LANA_ERR_CORRUPTION;
    *out = (uint64_t)value.as.number; return LANA_OK;
}

LanaError lana_ledger_open(LanaStore *store, LanaLedger **out) {
    LanaLedger *ledger;
    LanaError error;
    char *path;
    if (store == NULL || out == NULL) return LANA_ERR_INVALID_STATE;
    ledger = calloc(1u, sizeof(*ledger)); if (ledger == NULL) return LANA_ERR_OOM;
    ledger->store = store;

    if ((error = lana_store_get_path(store, &path)) != LANA_OK) {
        free(ledger); return error;
    }

    char journal_path[1024];
    (void)snprintf(journal_path, sizeof(journal_path), "%s/ledger.journal", path);
    free(path);

    ledger->journal = fopen(journal_path, "a+b");
    if (ledger->journal == NULL) {
        free(ledger); return LANA_ERR_IO;
    }

    *out = ledger; return LANA_OK;
}

LanaError lana_ledger_close(LanaLedger *ledger) {
    if (ledger == NULL) return LANA_OK;
    if (ledger->journal != NULL) fclose(ledger->journal);
    free(ledger); return LANA_OK;
}

LanaError lana_ledger_append(LanaLedger *ledger, const LanaEventInput *input, LanaEvent *out_event) {
    LanaStoreRevisionInfo revision; char key[64], json[2048]; Value value; LanaError error;
    if (ledger == NULL || input == NULL || input->struct_size < sizeof(*input) || input->schema_version != 1u ||
        !text_ok(input->entity) || !text_ok(input->actor) || !text_ok(input->action)) return LANA_ERR_SCHEMA;
    if (input->reason != NULL && strchr(input->reason, '"') != NULL) return LANA_ERR_SCHEMA;
    if (lana_store_current_revision(ledger->store, &revision) != LANA_OK || revision.revision_id == UINT64_MAX) return LANA_ERR_IO;
    (void)snprintf(key, sizeof(key), "event/%llu", (unsigned long long)(revision.revision_id + 1u));
    (void)snprintf(json, sizeof(json), "{\"action\":\"%s\",\"actor\":\"%s\",\"correction_of\":%llu,\"entity\":\"%s\",\"reason\":\"%s\",\"timestamp\":%llu}",
                   input->action, input->actor, (unsigned long long)input->correction_of, input->entity,
                   input->reason == NULL ? "" : input->reason, (unsigned long long)input->timestamp);
    if (strlen(json) >= sizeof(json) - 1u) return LANA_ERR_LIMIT;

    /* Write to separate append-only journal first. */
    if (ledger->journal != NULL) {
        if (fprintf(ledger->journal, "%s\n", json) < 0 || fflush(ledger->journal) != 0) return LANA_ERR_IO;
    }

    value = lana_value_string(json);
    if ((error = lana_store_put(ledger->store, key, value)) != LANA_OK) return error;
    if ((error = lana_store_commit(ledger->store, &revision)) != LANA_OK) return error;
    if (out_event != NULL) {
        memset(out_event, 0, sizeof(*out_event)); out_event->struct_size = sizeof(*out_event); out_event->schema_version = 1u;
        out_event->event_id = revision.revision_id; out_event->revision = revision.revision_id;
        out_event->entity = input->entity; out_event->actor = input->actor; out_event->action = input->action;
        out_event->reason = input->reason == NULL ? "" : input->reason; out_event->timestamp = input->timestamp;
        out_event->correction_of = input->correction_of;
    }
    return LANA_OK;
}

/* Scan the event journal, applying filters. Returns LANA_OK with count possibly
 * zero, or a distinct corruption/I/O error. */
static LanaError scan_events(LanaLedger *ledger, LanaVM *vm,
                             const char *entity, const char *actor, const char *action,
                             uint64_t start_timestamp, uint64_t end_timestamp,
                             LanaEvent **out_events, size_t *out_count) {
    LanaStoreRevisionInfo revision; LanaEvent *events; size_t count = 0u, index;
    if (lana_store_current_revision(ledger->store, &revision) != LANA_OK) return LANA_ERR_IO;
    events = calloc((size_t)revision.revision_id, sizeof(*events));
    if (revision.revision_id != 0u && events == NULL) return LANA_ERR_OOM;
    for (index = 1u; index <= revision.revision_id; ++index) {
        char key[64]; Value raw, parsed; LanaEvent event; const char *text; LanaError get_error;
        (void)snprintf(key, sizeof(key), "event/%llu", (unsigned long long)index);
        get_error = lana_store_get(ledger->store, vm, key, &raw);
        if (get_error == LANA_ERR_NOT_FOUND) continue;
        if (get_error != LANA_OK) { free(events); return get_error; }
        if (raw.type != VAL_STRING) { free(events); return LANA_ERR_CORRUPTION; }
        if (lana_json_parse(vm, raw.as.string, &parsed) != LANA_OK || parsed.type != VAL_MAP) { free(events); return LANA_ERR_CORRUPTION; }
        memset(&event, 0, sizeof(event)); event.struct_size = sizeof(event); event.schema_version = 1u; event.event_id = index; event.revision = index;
        if (map_text(parsed.as.map, "entity", &text) != LANA_OK) { free(events); return LANA_ERR_CORRUPTION; } event.entity = text;
        if (map_text(parsed.as.map, "actor", &text) != LANA_OK) { free(events); return LANA_ERR_CORRUPTION; } event.actor = text;
        if (map_text(parsed.as.map, "action", &text) != LANA_OK) { free(events); return LANA_ERR_CORRUPTION; } event.action = text;
        if (map_text(parsed.as.map, "reason", &text) != LANA_OK) { free(events); return LANA_ERR_CORRUPTION; } event.reason = text;
        if (map_number(parsed.as.map, "timestamp", &event.timestamp) != LANA_OK) { free(events); return LANA_ERR_CORRUPTION; }
        if (entity != NULL && strcmp(entity, event.entity) != 0) continue;
        if (actor != NULL && strcmp(actor, event.actor) != 0) continue;
        if (action != NULL && strcmp(action, event.action) != 0) continue;
        if (start_timestamp != 0u && event.timestamp < start_timestamp) continue;
        if (end_timestamp != 0u && event.timestamp > end_timestamp) continue;
        events[count++] = event;
    }
    *out_events = events; *out_count = count; return LANA_OK;
}

LanaError lana_ledger_query(LanaLedger *ledger, LanaVM *vm, const LanaLedgerQuery *query,
                            LanaEvent **out_events, size_t *out_count) {
    LanaError error;
    if (ledger == NULL || vm == NULL || query == NULL || query->struct_size < sizeof(*query) || query->schema_version != 1u ||
        out_events == NULL || out_count == NULL || (query->start_timestamp > query->end_timestamp && query->end_timestamp != 0u)) return LANA_ERR_INVALID_STATE;
    error = scan_events(ledger, vm, query->entity, query->actor, query->action,
                        query->start_timestamp, query->end_timestamp, out_events, out_count);
    if (error != LANA_OK) return error;
    return *out_count == 0u ? LANA_ERR_NO_MATCHING_EVENT : LANA_OK;
}

void lana_ledger_events_free(LanaEvent *events, size_t count) { (void)count; free(events); }

LanaError lana_ledger_query_coverage(LanaLedger *ledger, LanaVM *vm,
                                     const LanaLedgerCoverageQuery *query,
                                     LanaLedgerCoverageEntry **out_entries,
                                     size_t *out_count) {
    LanaLedgerCoverageEntry *entries; size_t index;
    if (ledger == NULL || vm == NULL || query == NULL || query->struct_size < sizeof(*query) || query->schema_version != 1u ||
        out_entries == NULL || out_count == NULL || query->target_entities == NULL || query->target_count == 0u ||
        (query->start_timestamp > query->end_timestamp && query->end_timestamp != 0u)) return LANA_ERR_INVALID_STATE;
    entries = calloc(query->target_count, sizeof(*entries));
    if (entries == NULL) return LANA_ERR_OOM;
    for (index = 0u; index < query->target_count; ++index) {
        LanaEvent *events = NULL; size_t count = 0u; LanaError error;
        if (query->target_entities[index] == NULL) { free(entries); return LANA_ERR_SCHEMA; }
        error = scan_events(ledger, vm, query->target_entities[index], query->actor, query->action,
                            query->start_timestamp, query->end_timestamp, &events, &count);
        if (error != LANA_OK) { free(events); free(entries); return error; }
        entries[index].struct_size = sizeof(entries[index]); entries[index].schema_version = 1u;
        entries[index].entity = query->target_entities[index];
        entries[index].matched = count != 0u;
        entries[index].events = events; entries[index].event_count = count;
    }
    *out_entries = entries; *out_count = query->target_count; return LANA_OK;
}

void lana_ledger_coverage_free(LanaLedgerCoverageEntry *entries, size_t count) {
    size_t index;
    if (entries == NULL) return;
    for (index = 0u; index < count; ++index) free(entries[index].events);
    free(entries);
}

static int hex_value(char character) {
    if (character >= '0' && character <= '9') return character - '0';
    if (character >= 'a' && character <= 'f') return character - 'a' + 10;
    if (character >= 'A' && character <= 'F') return character - 'A' + 10;
    return -1;
}

static LanaError hex_decode(const char *text, unsigned char out[32]) {
    size_t index;
    if (text == NULL || strlen(text) != 64u) return LANA_ERR_CORRUPTION;
    for (index = 0u; index < 32u; ++index) {
        int high = hex_value(text[index * 2u]);
        int low = hex_value(text[index * 2u + 1u]);
        if (high < 0 || low < 0) return LANA_ERR_CORRUPTION;
        out[index] = (unsigned char)((high << 4) | low);
    }
    return LANA_OK;
}

/* Parse the trailing numeric id from a key of the form "prefix/{id}" or
 * "prefix/{id}/{attempt}". Returns 0 on malformed input. */
static uint64_t key_tail_id(const char *key) {
    const char *slash = strrchr(key, '/');
    if (slash == NULL || slash[1] == '\0') return 0u;
    return (uint64_t)strtoull(slash + 1, NULL, 10);
}

/* Stored records are JSON text (VAL_STRING); decode to a map. */
static LanaError scan_value_map(LanaVM *vm, const Value *value, Value *out) {
    if (value->type == VAL_MAP) { *out = *value; return LANA_OK; }
    if (value->type != VAL_STRING) return LANA_ERR_CORRUPTION;
    if (lana_json_parse(vm, value->as.string, out) != LANA_OK || out->type != VAL_MAP)
        return LANA_ERR_CORRUPTION;
    return LANA_OK;
}

LanaError lana_ledger_traverse(LanaLedger *ledger, LanaVM *vm,
                               const LanaTraversalQuery *query,
                               LanaTraversalRecord **out_records,
                               size_t *out_count) {
    LanaStoreScanRecord *decisions = NULL, *events = NULL, *attempts = NULL;
    size_t decision_count = 0u, event_count = 0u, attempt_count = 0u;
    LanaTraversalRecord *records = NULL; size_t count = 0u, capacity = 0u;
    size_t decision_index, event_index;
    LanaError error;
    if (ledger == NULL || vm == NULL || query == NULL || query->struct_size < sizeof(*query) || query->schema_version != 1u ||
        query->evidence_id == NULL || out_records == NULL || out_count == NULL) return LANA_ERR_INVALID_STATE;
    error = lana_store_scan(ledger->store, vm, "decision/", &decisions, &decision_count);
    if (error != LANA_OK) return error;
    error = lana_store_scan(ledger->store, vm, "event/", &events, &event_count);
    if (error != LANA_OK) goto done;
    for (decision_index = 0u; decision_index < decision_count; ++decision_index) {
        Value parsed; const char *text; uint64_t decision_id, outcome_value;
        LanaTraversalRecord record; char decision_text[32];
        if (scan_value_map(vm, &decisions[decision_index].value, &parsed) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        if (map_text(parsed.as.map, "evidence_ids", &text) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        if (strstr(text, query->evidence_id) == NULL) continue;
        memset(&record, 0, sizeof(record)); record.struct_size = sizeof(record); record.schema_version = 1u;
        if (map_number(parsed.as.map, "decision_id", &decision_id) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        record.decision_id = decision_id;
        if (map_text(parsed.as.map, "policy_id", &text) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        record.policy_id = strdup(text);
        if (record.policy_id == NULL) { error = LANA_ERR_OOM; goto done; }
        if (map_text(parsed.as.map, "policy_version", &text) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        if (hex_decode(text, record.policy_version) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        if (map_number(parsed.as.map, "outcome", &outcome_value) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        record.outcome = (PolicyOutcome)(unsigned)outcome_value;
        if (map_text(parsed.as.map, "effect", &text) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
        record.effect = strdup(text);
        if (record.effect == NULL) { error = LANA_ERR_OOM; goto done; }
        /* Link the effect attempt for this decision (lowest attempt id). */
        {
            char attempt_prefix[64];
            (void)snprintf(attempt_prefix, sizeof(attempt_prefix), "effect-attempt/%llu/",
                           (unsigned long long)decision_id);
            error = lana_store_scan(ledger->store, vm, attempt_prefix, &attempts, &attempt_count);
            if (error != LANA_OK) goto done;
            if (attempt_count != 0u) {
                Value attempt_parsed; uint64_t status_value;
                if (scan_value_map(vm, &attempts[0].value, &attempt_parsed) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
                if (map_number(attempt_parsed.as.map, "status", &status_value) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
                record.attempt_id = key_tail_id(attempts[0].key);
                record.attempt_status = (EffectStatus)(unsigned)status_value;
            }
            lana_store_scan_free(attempts, attempt_count); attempts = NULL; attempt_count = 0u;
        }
        /* Link the event referencing this decision. */
        (void)snprintf(decision_text, sizeof(decision_text), "%llu", (unsigned long long)decision_id);
        for (event_index = 0u; event_index < event_count; ++event_index) {
            Value event_parsed; const char *reason;
            if (scan_value_map(vm, &events[event_index].value, &event_parsed) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
            if (map_text(event_parsed.as.map, "reason", &reason) != LANA_OK) { error = LANA_ERR_CORRUPTION; goto done; }
            if (strstr(reason, decision_text) != NULL) { record.event_id = key_tail_id(events[event_index].key); break; }
        }
        if (count == capacity) {
            size_t new_capacity = capacity == 0u ? 4u : capacity * 2u;
            LanaTraversalRecord *grown = realloc(records, new_capacity * sizeof(*grown));
            if (grown == NULL) { error = LANA_ERR_OOM; goto done; }
            records = grown; capacity = new_capacity;
        }
        records[count++] = record;
    }
    if (count == 0u) error = LANA_ERR_NO_MATCHING_EVENT;
done:
    lana_store_scan_free(decisions, decision_count);
    lana_store_scan_free(events, event_count);
    lana_store_scan_free(attempts, attempt_count);
    if (error != LANA_OK) { lana_traversal_records_free(records, count); return error; }
    *out_records = records; *out_count = count; return LANA_OK;
}

void lana_traversal_records_free(LanaTraversalRecord *records, size_t count) {
    size_t index;
    if (records == NULL) return;
    for (index = 0u; index < count; ++index) {
        free((void *)records[index].policy_id);
        free((void *)records[index].effect);
    }
    free(records);
}

File: runtime/rust/lana-runtime/src/ledger.rs
//! Append-only event ledger, mirroring `runtime/c/ledger.c` and
//! `runtime/include/ledger.h`.

use std::fs::{File, OpenOptions};
use std::io::Write;
use std::sync::{Arc, Mutex};

use lana_bytecode::LanaError;
use lana_vm::value::{Map, Value, ValueKind};

use crate::data::json_parse;
use crate::effects::EffectStatus;
use crate::policy::PolicyOutcome;
use crate::store::{ScanRecord, Store, store_commit, store_current_revision, store_get, store_get_path, store_put, store_scan};

/// Input to `ledger_append`, mirroring `LanaEventInput`.
pub struct EventInput {
    pub schema_version: u32,
    pub entity: Arc<str>,
    pub actor: Arc<str>,
    pub action: Arc<str>,
    pub reason: Option<Arc<str>>,
    pub timestamp: u64,
    pub correction_of: u64,
}

/// A stored event, mirroring `LanaEvent`.
pub struct Event {
    pub schema_version: u32,
    pub event_id: u64,
    pub entity: Arc<str>,
    pub actor: Arc<str>,
    pub action: Arc<str>,
    pub reason: Arc<str>,
    pub timestamp: u64,
    pub revision: u64,
    pub correction_of: u64,
}

/// A query filter, mirroring `LanaLedgerQuery`.
pub struct LedgerQuery {
    pub schema_version: u32,
    pub entity: Option<Arc<str>>,
    pub actor: Option<Arc<str>>,
    pub action: Option<Arc<str>>,
    pub start_timestamp: u64,
    pub end_timestamp: u64,
}

/// A coverage query, mirroring `LanaLedgerCoverageQuery`.
pub struct CoverageQuery {
    pub schema_version: u32,
    pub entity: Option<Arc<str>>,
    pub actor: Option<Arc<str>>,
    pub action: Option<Arc<str>>,
    pub start_timestamp: u64,
    pub end_timestamp: u64,
    pub target_entities: Vec<Arc<str>>,
}

/// One coverage result, mirroring `LanaLedgerCoverageEntry`.
pub struct CoverageEntry {
    pub schema_version: u32,
    pub entity: Arc<str>,
    pub matched: bool,
    pub events: Vec<Event>,
}

/// An evidence-to-event traversal query, mirroring `LanaTraversalQuery`.
pub struct TraversalQuery {
    pub schema_version: u32,
    pub evidence_id: Arc<str>,
}

/// One traversal record, mirroring `LanaTraversalRecord`.
pub struct TraversalRecord {
    pub schema_version: u32,
    pub decision_id: u64,
    pub policy_id: Arc<str>,
    pub policy_version: [u8; 32],
    pub outcome: PolicyOutcome,
    pub effect: Arc<str>,
    pub attempt_id: u64,
    pub attempt_status: EffectStatus,
    pub event_id: u64,
}

/// The ledger, mirroring `struct LanaLedger`. Borrows the store it appends to.
pub struct Ledger<'a> {
    store: &'a mut Store,
    journal: Option<File>,
}

fn text_ok(text: &str) -> bool {
    !text.is_empty()
}

fn map_text(map: &Map, key: &str) -> Result<Arc<str>, LanaError> {
    match map.get(key) {
        Some(Value { kind: ValueKind::String(s), .. }) if !s.is_empty() => Ok(s.clone()),
        _ => Err(LanaError::Corruption),
    }
}

fn map_number(map: &Map, key: &str) -> Result<u64, LanaError> {
    match map.get(key) {
        Some(Value { kind: ValueKind::Number(n), .. }) => Ok(*n as u64),
        _ => Err(LanaError::Corruption),
    }
}

fn hex_value(character: u8) -> Option<u8> {
    match character {
        b'0'..=b'9' => Some(character - b'0'),
        b'a'..=b'f' => Some(character - b'a' + 10),
        b'A'..=b'F' => Some(character - b'A' + 10),
        _ => None,
    }
}

fn hex_decode(text: &str) -> Result<[u8; 32], LanaError> {
    if text.len() != 64 {
        return Err(LanaError::Corruption);
    }
    let bytes = text.as_bytes();
    let mut out = [0u8; 32];
    for (index, byte) in out.iter_mut().enumerate() {
        let high = hex_value(bytes[index * 2]).ok_or(LanaError::Corruption)?;
        let low = hex_value(bytes[index * 2 + 1]).ok_or(LanaError::Corruption)?;
        *byte = (high << 4) | low;
    }
    Ok(out)
}

fn key_tail_id(key: &str) -> u64 {
    match key.rfind('/') {
        Some(pos) if pos + 1 < key.len() => key[pos + 1..].parse().unwrap_or(0),
        _ => 0,
    }
}

/// Decode a stored record (JSON text or an already-parsed map) to a map.
fn scan_value_map(value: &Value) -> Result<Arc<Mutex<Map>>, LanaError> {
    match &value.kind {
        ValueKind::Map(map) => Ok(map.clone()),
        ValueKind::String(s) => {
            let parsed = json_parse(s)?;
            match parsed.kind {
                ValueKind::Map(map) => Ok(map),
                _ => Err(LanaError::Corruption),
            }
        }
        _ => Err(LanaError::Corruption),
    }
}

pub fn ledger_open(store: &mut Store) -> Result<Ledger<'_>, LanaError> {
    let path = store_get_path(store)?;
    let journal_path = format!("{}/ledger.journal", path);
    let journal = OpenOptions::new()
        .read(true)
        .append(true)
        .create(true)
        .open(&journal_path)
        .map_err(|_| LanaError::Io)?;
    Ok(Ledger { store, journal: Some(journal) })
}

pub fn ledger_close(ledger: &mut Ledger<'_>) -> Result<(), LanaError> {
    ledger.journal = None;
    Ok(())
}

pub fn ledger_append(ledger: &mut Ledger<'_>, input: &EventInput) -> Result<Event, LanaError> {
    if input.schema_version != 1
        || !text_ok(&input.entity)
        || !text_ok(&input.actor)
        || !text_ok(&input.action)
    {
        return Err(LanaError::Schema);
    }
    if input.reason.as_deref().map_or(false, |r| r.contains('"')) {
        return Err(LanaError::Schema);
    }
    let revision = store_current_revision(ledger.store)?;
    if revision.revision_id == u64::MAX {
        return Err(LanaError::Io);
    }
    let key = format!("event/{}", revision.revision_id + 1);
    let json = format!(
        "{{\"action\":\"{}\",\"actor\":\"{}\",\"correction_of\":{},\"entity\":\"{}\",\"reason\":\"{}\",\"timestamp\":{}}}",
        input.action,
        input.actor,
        input.correction_of,
        input.entity,
        input.reason.as_deref().unwrap_or(""),
        input.timestamp
    );
    if let Some(journal) = ledger.journal.as_mut() {
        writeln!(journal, "{}", json).map_err(|_| LanaError::Io)?;
        journal.flush().map_err(|_| LanaError::Io)?;
    }
    store_put(ledger.store, &key, &Value::string(Arc::from(json.as_str())))?;
    let committed = store_commit(ledger.store)?;
    Ok(Event {
        schema_version: 1,
        event_id: committed.revision_id,
        entity: input.entity.clone(),
        actor: input.actor.clone(),
        action: input.action.clone(),
        reason: input.reason.clone().unwrap_or_else(|| Arc::from("")),
        timestamp: input.timestamp,
        revision: committed.revision_id,
        correction_of: input.correction_of,
    })
}

fn scan_events(
    store: &Store,
    entity: Option<&str>,
    actor: Option<&str>,
    action: Option<&str>,
    start_timestamp: u64,
    end_timestamp: u64,
) -> Result<Vec<Event>, LanaError> {
    let revision = store_current_revision(store)?;
    let mut events = Vec::new();
    for index in 1..=revision.revision_id {
        let key = format!("event/{}", index);
        let raw = match store_get(store, &key) {
            Ok(raw) => raw,
            Err(LanaError::NotFound) => continue,
            Err(e) => return Err(e),
        };
        let map = match &raw.kind {
            ValueKind::String(s) => {
                let parsed = json_parse(s)?;
                match parsed.kind {
                    ValueKind::Map(map) => map,
                    _ => return Err(LanaError::Corruption),
                }
            }
            _ => return Err(LanaError::Corruption),
        };
        let map = map.lock().unwrap();
        let event_entity = map_text(&map, "entity")?;
        let event_actor = map_text(&map, "actor")?;
        let event_action = map_text(&map, "action")?;
        let event_reason = map_text(&map, "reason")?;
        let timestamp = map_number(&map, "timestamp")?;
        if entity.map_or(false, |e| e != &*event_entity) {
            continue;
        }
        if actor.map_or(false, |a| a != &*event_actor) {
            continue;
        }
        if action.map_or(false, |a| a != &*event_action) {
            continue;
        }
        if start_timestamp != 0 && timestamp < start_timestamp {
            continue;
        }
        if end_timestamp != 0 && timestamp > end_timestamp {
            continue;
        }
        events.push(Event {
            schema_version: 1,
            event_id: index,
            entity: event_entity,
            actor: event_actor,
            action: event_action,
            reason: event_reason,
            timestamp,
            revision: index,
            correction_of: 0,
        });
    }
    Ok(events)
}

pub fn ledger_query(ledger: &Ledger<'_>, query: &LedgerQuery) -> Result<Vec<Event>, LanaError> {
    if query.schema_version != 1
        || (query.start_timestamp > query.end_timestamp && query.end_timestamp != 0)
    {
        return Err(LanaError::InvalidState);
    }
    let events = scan_events(
        ledger.store,
        query.entity.as_deref(),
        query.actor.as_deref(),
        query.action.as_deref(),
        query.start_timestamp,
        query.end_timestamp,
    )?;
    if events.is_empty() {
        return Err(LanaError::NoMatchingEvent);
    }
    Ok(events)
}

pub fn ledger_query_coverage(
    ledger: &Ledger<'_>,
    query: &CoverageQuery,
) -> Result<Vec<CoverageEntry>, LanaError> {
    if query.schema_version != 1
        || query.target_entities.is_empty()
        || (query.start_timestamp > query.end_timestamp && query.end_timestamp != 0)
    {
        return Err(LanaError::InvalidState);
    }
    let mut entries = Vec::with_capacity(query.target_entities.len());
    for target in &query.target_entities {
        if target.is_empty() {
            return Err(LanaError::Schema);
        }
        let events = scan_events(
            ledger.store,
            Some(target),
            query.actor.as_deref(),
            query.action.as_deref(),
            query.start_timestamp,
            query.end_timestamp,
        )?;
        entries.push(CoverageEntry {
            schema_version: 1,
            entity: target.clone(),
            matched: !events.is_empty(),
            events,
        });
    }
    Ok(entries)
}

pub fn ledger_traverse(
    ledger: &Ledger<'_>,
    query: &TraversalQuery,
) -> Result<Vec<TraversalRecord>, LanaError> {
    if query.schema_version != 1 || query.evidence_id.is_empty() {
        return Err(LanaError::InvalidState);
    }
    let decisions: Vec<ScanRecord> = store_scan(ledger.store, "decision/")?;
    let events: Vec<ScanRecord> = store_scan(ledger.store, "event/")?;
    let mut records: Vec<TraversalRecord> = Vec::new();
    for decision in &decisions {
        let parsed = scan_value_map(&decision.value)?;
        let parsed = parsed.lock().unwrap();
        let evidence_ids = map_text(&parsed, "evidence_ids")?;
        if !evidence_ids.contains(&*query.evidence_id) {
            continue;
        }
        let decision_id = map_number(&parsed, "decision_id")?;
        let policy_id = map_text(&parsed, "policy_id")?;
        let policy_version_text = map_text(&parsed, "policy_version")?;
        let policy_version = hex_decode(&policy_version_text)?;
        let outcome_value = map_number(&parsed, "outcome")?;
        let effect = map_text(&parsed, "effect")?;
        drop(parsed);

        let attempt_prefix = format!("effect-attempt/{}/", decision_id);
        let attempts: Vec<ScanRecord> = store_scan(ledger.store, &attempt_prefix)?;
        let mut attempt_id = 0u64;
        let mut attempt_status = EffectStatus::Pending;
        if let Some(first) = attempts.first() {
            let attempt_parsed = scan_value_map(&first.value)?;
            let attempt_parsed = attempt_parsed.lock().unwrap();
            let status_value = map_number(&attempt_parsed, "status")?;
            attempt_id = key_tail_id(&first.key);
            attempt_status = match status_value {
                0 => EffectStatus::Pending,
                1 => EffectStatus::Succeeded,
                2 => EffectStatus::Failed,
                3 => EffectStatus::Cancelled,
                _ => return Err(LanaError::Corruption),
            };
        }

        let decision_text = decision_id.to_string();
        let mut event_id = 0u64;
        for event in &events {
            let event_parsed = scan_value_map(&event.value)?;
            let event_parsed = event_parsed.lock().unwrap();
            let reason = map_text(&event_parsed, "reason")?;
            if reason.contains(&decision_text) {
                event_id = key_tail_id(&event.key);
                break;
            }
        }

        records.push(TraversalRecord {
            schema_version: 1,
            decision_id,
            policy_id,
            policy_version,
            outcome: match outcome_value {
                0 => PolicyOutcome::Authorize,
                1 => PolicyOutcome::Refuse,
                2 => PolicyOutcome::RequestMoreEvidence,
                _ => return Err(LanaError::Corruption),
            },
            effect,
            attempt_id,
            attempt_status,
            event_id,
        });
    }
    if records.is_empty() {
        return Err(LanaError::NoMatchingEvent);
    }
    Ok(records)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn temp_store(name: &str) -> String {
        let dir = std::env::temp_dir().join(format!("lana_ledger_{}_{}", name, std::process::id()));
        let _ = std::fs::remove_dir_all(&dir);
        dir.to_str().unwrap().to_string()
    }

    #[test]
    fn append_and_query() {
        let path = temp_store("append");
        let mut store = crate::store::store_open(&crate::store::StoreOptions {
            schema_version: 1,
            path: path.clone(),
            timeout_ms: 0,
        })
        .unwrap();
        let mut ledger = ledger_open(&mut store).unwrap();
        let event = ledger_append(
            &mut ledger,
            &EventInput {
                schema_version: 1,
                entity: Arc::from("e1"),
                actor: Arc::from("a1"),
                action: Arc::from("grant"),
                reason: Some(Arc::from("approved")),
                timestamp: 100,
                correction_of: 0,
            },
        )
        .unwrap();
        assert_eq!(event.event_id, 1);
        let events = ledger_query(
            &ledger,
            &LedgerQuery {
                schema_version: 1,
                entity: Some(Arc::from("e1")),
                actor: None,
                action: None,
                start_timestamp: 0,
                end_timestamp: 0,
            },
        )
        .unwrap();
        assert_eq!(events.len(), 1);
        assert_eq!(events[0].action, Arc::from("grant"));
        ledger_close(&mut ledger).unwrap();
        let _ = std::fs::remove_dir_all(&path);
    }
}
